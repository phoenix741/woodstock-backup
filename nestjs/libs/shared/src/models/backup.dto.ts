import { Field, ObjectType } from '@nestjs/graphql';
import { Transform } from 'class-transformer';

@ObjectType()
export class Backup {
  number!: number;
  completed!: boolean;

  startDate!: number;
  endDate?: number;

  fileCount!: number;
  newFileCount!: number;
  existingFileCount!: number;
  removedFileCount!: number;
  modifiedFileCount!: number;

  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  fileSize!: bigint;
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  existingFileSize!: bigint;
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  newFileSize!: bigint;
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  modifiedFileSize!: bigint;

  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  compressedFileSize!: bigint;
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  existingCompressedFileSize!: bigint;
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  newCompressedFileSize!: bigint;
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  modifiedCompressedFileSize!: bigint;

  speed!: number;
}
