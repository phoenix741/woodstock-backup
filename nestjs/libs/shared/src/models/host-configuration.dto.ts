import { createUnionType, Field, ObjectType } from '@nestjs/graphql';
import { ApiExtraModels, ApiProperty, getSchemaPath } from '@nestjs/swagger';
import { Exclude } from 'class-transformer';
import { IsNumber, Matches, Max, Min, ValidateNested } from 'class-validator';
import { Schedule } from './scheduler.dto';

/**
 * Part of config file.
 *
 * Store information about a share
 */
@ObjectType()
export class BackupTaskShare {
  @ApiProperty({ example: '/home' })
  name!: string;

  @ApiProperty({ example: ['/home/user'] })
  includes?: string[];

  @ApiProperty({ example: ['*.bak'] })
  excludes?: string[];
}

/**
 * Part of config file
 *
 * Store information about a DHCP Address
 */
@ObjectType()
export class DhcpAddress {
  @ApiProperty({ example: '192.168.101' })
  @Matches(/^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){2}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$/)
  address!: string;

  @ApiProperty({ example: 0 })
  @Min(0)
  @Max(255)
  start!: number;

  @ApiProperty({ example: 50 })
  @Min(0)
  @Max(255)
  end!: number;
}

@ObjectType()
export class ExecuteCommandOperation {
  @ApiProperty({ type: String, enum: ['ExecuteCommand'] })
  @Field(() => String)
  name!: 'ExecuteCommand';

  @ApiProperty({ example: '/bin/true' })
  command!: string;
}

@ObjectType()
export class BackupOperation {
  @ApiProperty({ type: String, enum: ['Backup'] })
  @Field(() => String)
  name!: 'Backup';

  @ValidateNested()
  shares!: Array<BackupTaskShare>;

  @ApiProperty({ example: [] })
  includes?: Array<string>;

  @ApiProperty({ example: [] })
  excludes?: Array<string>;

  @ApiProperty({ example: 1200 })
  @IsNumber()
  timeout?: number;
}

export type Operation = ExecuteCommandOperation | BackupOperation;

export const OperationUnion = createUnionType({
  name: 'Operation',
  types: () => [ExecuteCommandOperation, BackupOperation],
  resolveType(value) {
    if (value.command) {
      return ExecuteCommandOperation;
    } else if (value.share) {
      return BackupOperation;
    }
    return null;
  },
});

@ApiExtraModels(BackupOperation, ExecuteCommandOperation)
@ObjectType()
export class HostConfigOperation {
  @ApiProperty({
    type: 'array',
    items: {
      oneOf: [{ $ref: getSchemaPath(ExecuteCommandOperation) }, { $ref: getSchemaPath(BackupOperation) }],
    },
  })
  @ValidateNested()
  @Field(() => [OperationUnion])
  tasks?: Operation[];

  @ApiProperty({
    type: 'array',
    items: {
      oneOf: [{ $ref: getSchemaPath(ExecuteCommandOperation) }, { $ref: getSchemaPath(BackupOperation) }],
    },
  })
  @ValidateNested()
  @Field(() => [OperationUnion])
  finalizeTasks?: Operation[];
}

/**
 * Config file for one Host
 *
 * Contains all information that can be used to backup a host.
 */
@ObjectType()
export class HostConfiguration {
  @ApiProperty({ type: Boolean })
  @Field(() => Boolean)
  isLocal?: boolean;

  @Exclude()
  password: string;

  @ApiProperty({ example: [] })
  addresses?: string[];

  @ValidateNested()
  dhcp?: DhcpAddress[];

  @ValidateNested()
  operations?: HostConfigOperation = new HostConfigOperation();

  @ValidateNested()
  schedule?: Schedule;
}
