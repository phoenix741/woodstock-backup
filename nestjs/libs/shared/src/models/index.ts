export * from './backup.dto.js';
export * from './backups-files.dto.js';
export * from './host-configuration.dto.js';
export * from './object-proto.model.js';
export * from './pool-chunk.dto.js';
export * from './scheduler.dto.js';
export * from './stats.js';
export * from './stats.dto.js';
export * from './stats.model.js';
export * from './tasks.dto.js';
export * from './woodstock.js';
export * from './woodstockClient.js';
