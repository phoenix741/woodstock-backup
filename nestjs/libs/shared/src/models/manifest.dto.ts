import { FileManifest } from './woodstock';

export interface ManifestChunk {
  sha256: Buffer;
  manifest: FileManifest;
}
