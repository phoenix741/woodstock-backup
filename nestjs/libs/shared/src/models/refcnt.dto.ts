export interface RefcntJobData {
  host?: string;
  number?: number;

  fix?: boolean;
  refcnt?: boolean;
  unused?: boolean;

  target?: string;
}
