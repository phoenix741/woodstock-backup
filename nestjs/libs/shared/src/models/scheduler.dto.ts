import { ObjectType } from '@nestjs/graphql';

@ObjectType()
export class ScheduledBackupToKeep {
  hourly?: number;
  daily?: number;
  weekly?: number;
  monthly?: number;
  yearly?: number;
}

@ObjectType()
export class Schedule {
  activated?: boolean;
  backupPeriod?: number;
  backupToKeep?: ScheduledBackupToKeep;
}

export class ApplicationScheduler {
  wakeupSchedule?: string;
  nightlySchedule?: string;
  defaultSchedule?: Schedule = new Schedule();

  constructor(s: Partial<ApplicationScheduler> = {}) {
    Object.assign(this, s);
  }
}
