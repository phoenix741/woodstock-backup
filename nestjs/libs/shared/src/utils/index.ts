export * from './concurrent-promise.util';
export * from './global-regexp.utils';
export * from './iterator.utils';
export * from './number.utils';
export * from './objects.utils';
export * from './path.utils';
export * from './rendering.utils';
