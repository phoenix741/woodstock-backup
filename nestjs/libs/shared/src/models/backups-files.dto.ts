import { Field, HideField, Int, ObjectType, registerEnumType } from '@nestjs/graphql';
import { ApiHideProperty, ApiProperty } from '@nestjs/swagger';
import {
  JsFileManifest,
  JsFileManifestAcl,
  JsFileManifestAclQualifier,
  JsFileManifestStat,
  JsFileManifestType,
  JsFileManifestXAttr,
} from '@woodstock/shared-rs';
import { Exclude, Expose, Transform } from 'class-transformer';

import { mangle } from '../utils';

export enum EnumFileType {
  RegularFile = 0,
  Symlink = 1,
  Directory = 2,
  BlockDevice = 3,
  CharacterDevice = 4,
  Fifo = 5,
  Socket = 6,
  Unknown = 99,
}

export enum FileManifestAclQualifier {
  Undefined = 0,
  UserObj = 1,
  GroupObj = 2,
  Other = 3,
  UserId = 4,
  GroupId = 5,
  Mask = 6,
}

registerEnumType(EnumFileType, {
  name: 'EnumFileType',
});

registerEnumType(FileManifestAclQualifier, {
  name: 'FileManifestAclQualifier',
});

@ObjectType()
export class FileStat {
  @ApiProperty({ type: () => Number })
  @Field(() => Int)
  ownerId: number;

  @ApiProperty({ type: () => Number })
  @Field(() => Int)
  groupId: number;

  @Transform(({ value }) => value?.toString())
  @ApiProperty({ type: () => String })
  @Field(() => String)
  size: bigint;

  @Transform(({ value }) => value?.toString())
  @ApiProperty({ type: () => String })
  @Field(() => String)
  compressedSize: bigint;

  @ApiProperty({ type: () => Number })
  @Field(() => String)
  lastRead: number;

  @ApiProperty({ type: () => Number })
  @Field(() => String)
  lastModified: number;

  @ApiProperty({ type: () => Number })
  @Field(() => String)
  created: number;

  @ApiProperty({ type: () => Number })
  @Field(() => Int)
  mode: number;

  @ApiProperty({ type: () => Number, enum: EnumFileType })
  @Field(() => EnumFileType)
  type: JsFileManifestType;

  @ApiProperty({ type: () => String })
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  dev: bigint;

  @ApiProperty({ type: () => String })
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  rdev: bigint;

  @ApiProperty({ type: () => String })
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  ino: bigint;

  @ApiProperty({ type: () => String })
  @Transform((v) => BigInt(v.value))
  @Field(() => BigInt)
  nlink: bigint;

  constructor(partial: Partial<JsFileManifestStat>) {
    Object.assign(this, partial);
  }
}

@ObjectType()
export class FileAcl implements JsFileManifestAcl {
  @ApiProperty({ type: () => FileManifestAclQualifier })
  @Field(() => FileManifestAclQualifier)
  qualifier: JsFileManifestAclQualifier;
  @ApiProperty({ type: () => Number })
  @Field(() => Int)
  id: number;
  @ApiProperty({ type: () => Number })
  @Field(() => Int)
  perm: number;

  constructor(partial: Partial<JsFileManifestAcl>) {
    Object.assign(this, partial);
  }
}

@ObjectType()
export class FileXattr implements JsFileManifestXAttr {
  @Transform(({ value }) => mangle(value))
  @ApiProperty({ type: () => String })
  @Field(() => String)
  key: Buffer;

  @Transform(({ value }) => mangle(value))
  @ApiProperty({ type: () => String })
  @Field(() => String)
  value: Buffer;

  constructor(partial: Partial<JsFileManifestXAttr>) {
    Object.assign(this, partial);
  }
}

@ObjectType()
export class FileDescription {
  @Transform(({ value }) => mangle(value))
  @ApiProperty({ type: () => String })
  @Field(() => String)
  path: Buffer;

  stats: FileStat | undefined;

  @Field(() => FileXattr)
  xattr: FileXattr[];

  @Field(() => FileAcl)
  acl: FileAcl[];

  @Transform(({ value }) => value && mangle(value))
  @ApiProperty({ type: () => String })
  @Field(() => String)
  symlink: Buffer;

  @Exclude()
  @HideField()
  @ApiHideProperty()
  chunks: Buffer[];

  @Exclude()
  @HideField()
  @ApiHideProperty()
  hash: Buffer;

  @Exclude()
  @HideField()
  @ApiHideProperty()
  metadata: Record<string, Buffer>;

  constructor(partial: JsFileManifest) {
    Object.assign(this, partial);
    this.stats = this.stats && new FileStat(this.stats);
    this.acl = this.acl && this.acl.map((acl) => new FileAcl(acl));
  }

  @Expose()
  @ApiProperty({ type: Number, enum: EnumFileType })
  @Field(() => EnumFileType)
  get type(): JsFileManifestType {
    return this.stats?.type ?? JsFileManifestType.Unknown;
  }
}
