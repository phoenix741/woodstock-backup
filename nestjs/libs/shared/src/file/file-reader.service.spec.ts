import { Test, TestingModule } from '@nestjs/testing';
import { AsyncIterableX, from, toArray } from 'ix/asynciterable';
import * as Long from 'long';
import { IndexManifest } from '../manifest/index-manifest.model';
import { FileManifest } from '../models/woodstock';
import { globStringToRegex } from '../utils/global-regexp.utils';
import { FileBrowserService } from './file-browser.service';
import { FileReader } from './file-reader.service';

describe('FileReader', () => {
  let service: FileReader;

  const fakeFileBrowserService = {
    getFiles() {
      return (): AsyncIterableX<FileManifest> => {
        return from([
          {
            path: Buffer.from('/file-reader.service.ts'),
            stats: {
              lastModified: Long.fromValue(122),
              size: Long.fromValue(123),
              mode: Long.fromValue(0o0100000),
            },
          } as FileManifest,
        ]);
      };
    },
  };

  beforeEach(async () => {
    const app: TestingModule = await Test.createTestingModule({
      providers: [FileReader, { provide: FileBrowserService, useValue: fakeFileBrowserService }],
    }).compile();

    service = app.get<FileReader>(FileReader);
  });

  test('Calculate hash of the files', async () => {
    const index = new IndexManifest();
    index.add({
      path: Buffer.from('/file-reader.service.ts'),
      stats: {
        lastModified: Long.fromValue(123),
        size: Long.fromValue(124),
        mode: Long.fromValue(0o0100000),
      },
      acl: [],
      xattr: {},
      chunks: [],
    });

    const value = await toArray(service.getFiles(index, Buffer.from(__dirname), [], [globStringToRegex('*.spec.ts')]));

    expect(value).toMatchSnapshot();
  });
});
