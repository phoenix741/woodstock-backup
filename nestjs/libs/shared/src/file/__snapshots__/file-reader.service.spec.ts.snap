// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileReader Calculate hash of the files 1`] = `
Array [
  Object {
    "manifest": Object {
      "chunks": Array [
        Object {
          "data": Array [
            17,
            18,
            245,
            27,
            213,
            27,
            188,
            176,
            136,
            235,
            142,
            68,
            201,
            58,
            195,
            206,
            162,
            199,
            89,
            197,
            56,
            253,
            161,
            43,
            22,
            228,
            121,
            14,
            243,
            254,
            36,
            62,
          ],
          "type": "Buffer",
        },
      ],
      "path": Object {
        "data": Array [
          47,
          102,
          105,
          108,
          101,
          45,
          114,
          101,
          97,
          100,
          101,
          114,
          46,
          115,
          101,
          114,
          118,
          105,
          99,
          101,
          46,
          116,
          115,
        ],
        "type": "Buffer",
      },
      "sha256": Object {
        "data": Array [
          17,
          18,
          245,
          27,
          213,
          27,
          188,
          176,
          136,
          235,
          142,
          68,
          201,
          58,
          195,
          206,
          162,
          199,
          89,
          197,
          56,
          253,
          161,
          43,
          22,
          228,
          121,
          14,
          243,
          254,
          36,
          62,
        ],
        "type": "Buffer",
      },
      "stats": Object {
        "lastModified": Long {
          "high": 0,
          "low": 122,
          "unsigned": false,
        },
        "mode": Long {
          "high": 0,
          "low": 32768,
          "unsigned": false,
        },
        "size": Long {
          "high": 0,
          "low": 123,
          "unsigned": false,
        },
      },
    },
    "type": 1,
  },
]
`;
